File: src/Native/C++/Omnix.Algorithms/avx2/Crc32_Castagnoli.cpp
#include "stdafx.h"
#include "Crc32_Castagnoli.h"

Crc32_Castagnoli::Crc32_Castagnoli()
{
    const uint32_t poly = 0x82F63B78;

    for (uint32_t i = 0; i < 256; i++)
    {
        uint32_t x = i;

        for (uint32_t j = 0; j < 8; j++)
        {
            if ((x & 1) != 0)
            {
                x = (x >> 1) ^ poly;
            }
            else
            {
                x >>= 1;
            }
        }

        _table[i] = x;
    }
}

Crc32_Castagnoli::~Crc32_Castagnoli()
{

}

uint32_t Crc32_Castagnoli::Compute(const uint32_t x, const byte* source, const int32_t length) const
{
    int v = x;

    for (int32_t i = 0; i < length; i++)
    {
        v = (v >> 8) ^ _table[((byte)(v & 0xff)) ^ source[i]];
    }

    return v;
}

uint32_t Crc32_Castagnoli_Compute(const uint32_t x, const byte* source, const int32_t length)
{
    return _crc32_castagnoli.Compute(x, source, length);
}

File: src/Omnius.Core.Serialization/Internal/Crc32_Castagnoli.cs
namespace Omnius.Core.Serialization;

public unsafe ref struct Crc32_Castagnoli
{
    private static readonly uint[] _table;

    static Crc32_Castagnoli()
    {
        // uint poly = 0xedb88320;
        uint poly = 0x82F63B78;
        _table = new uint[256];

        for (uint i = 0; i < 256; i++)
        {
            uint x = i;

            for (int j = 0; j < 8; j++)
            {
                if ((x & 1) != 0)
                {
                    x = (x >> 1) ^ poly;
                }
                else
                {
                    x >>= 1;
                }
            }

            _table[i] = x;
        }
    }

    private uint _result = 0xFFFFFFFF;

    public Crc32_Castagnoli()
    {
    }

    public void Compute(ReadOnlySpan<byte> memory)
    {
        fixed (uint* p_table = _table)
        fixed (byte* p_src = memory)
        {
            var t_src = p_src;

            for (int i = 0; i < memory.Length; i++)
            {
                _result = (_result >> 8) ^ p_table[(_result & 0xff) ^ *t_src++];
            }
        }
    }

    public uint GetResult()
    {
        return (uint)(_result ^ 0xFFFFFFFF);
    }
}
